runtime/runtime-core/common/src/aws/smithy/kotlin/runtime/collections/LruCache.kt

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package aws.smithy.kotlin.runtime.collections
+
+import aws.smithy.kotlin.runtime.InternalApi
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/**
+ * A thread-safe generic LRU (least recently used) cache.
+ * Entries will be added up to a configured [capacity].
+ * Once full, adding a new entry will evict the least recently used entry.
+ */
+@InternalApi
+public class LruCache<K, V>(
+    public val capacity: Int,
+) {
+    private val mu = Mutex() // protects map
+    private val map = linkedMapOf<K, V>()
+
+    /**
+     * Returns the value for a key [k], or null if it does not exist.
+     * @param k the key to look up
+     * @return the value associated with the key, or null if it does not exist
+     */
+    public suspend fun get(k: K): V? = mu.withLock {
+        map.moveKeyToBack(k)
+        return map[k]
+    }
+
+    /**
+     * Add or update a cache entry with a key [k] and value [v].
+     * @param k the key to associate the value with
+     * @param v the value to store in the cache
+     * @return [Unit]
+     */
+    public suspend fun put(k: K, v: V): Unit = mu.withLock {
+        if (k !in map && map.size == capacity) {
+            map.remove(map.keys.first())
```

Comment:
On JVM, [LinkedHashMap](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-linked-hash-map/) is typealiased to [Java's LinkedHashMap](https://docs.oracle.com/javase/8/docs/api/java/util/LinkedHashMap.html), which provides constant-time access to the first and last keys since it's backed by a doubly linked list